app/controllers/tradingNames/HasTradingNameController.scala

```diff
@@ -0,0 +1,98 @@
+import forms.tradingNames.HasTradingNameFormProvider
+import logging.Logging
+import models.requests.AuthenticatedDataRequest
+import pages.tradingNames.HasTradingNamePage
+import pages.{JourneyRecoveryPage, Waypoints}
+import play.api.data.Form
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents, Result}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import utils.FutureSyntax.FutureOps
+import views.html.tradingNames.HasTradingNameView
+
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class HasTradingNameController @Inject()(
+                                          override val messagesApi: MessagesApi,
+                                          cc: AuthenticatedControllerComponents,
+                                          formProvider: HasTradingNameFormProvider,
+                                          view: HasTradingNameView
+                                        )(implicit ec: ExecutionContext) extends FrontendBaseController with I18nSupport with Logging {
+
+  protected val controllerComponents: MessagesControllerComponents = cc
+
+  val form: Form[Boolean] = formProvider()
+
+  def onPageLoad(waypoints: Waypoints): Action[AnyContent] = cc.authAndGetData().async {
+    implicit request =>
+      getCompanyName(waypoints) {
+        companyName =>
+
+          val preparedForm = request.userAnswers.get(HasTradingNamePage) match {
+            case None => form
+            case Some(value) => form.fill(value)
+          }
+
+          Ok(view(preparedForm, waypoints, companyName)).toFuture
+      }
+  }
+
+  def onSubmit(waypoints: Waypoints): Action[AnyContent] = cc.authAndGetData().async {
+    implicit request =>
+      getCompanyName(waypoints) {
+        companyName =>
+
+          form.bindFromRequest().fold(
+            formWithErrors =>
+              BadRequest(view(formWithErrors, waypoints, companyName)).toFuture,
+
+            value =>
+              for {
+                updatedAnswers <- Future.fromTry(request.userAnswers.set(HasTradingNamePage, value))
+                _ <- cc.sessionRepository.set(updatedAnswers)
+              } yield Redirect(HasTradingNamePage.navigate(waypoints, request.userAnswers, updatedAnswers).route)
+          )
+      }
+  }
+
+  private def getCompanyName(waypoints: Waypoints)(block: String => Future[Result])
+                            (implicit request: AuthenticatedDataRequest[AnyContent]): Future[Result] = {
+    request.userAnswers.vatInfo match {
+      case Some(vatInfo) if vatInfo.organisationName.isDefined =>
+        val name = vatInfo.organisationName.getOrElse {
+          val exception = new IllegalStateException("No organisation name when expecting one")
+          logger.error(exception.getMessage, exception)
+          throw exception
+        }
+        block(name)
+      case Some(vatInfo) if vatInfo.individualName.isDefined =>
```

Comment:
The deferment to individualName if organisationName does not exist is not tested 

![image](https://github.com/hmrc/ioss-registration-frontend/assets/3098723/24ac6f69-ce6c-415f-8f75-234cf1d52f9c)

Reply:
Added additional test to cover individual name